### PR TITLE
Implemented YoloV5 as a 3rd model for inference.

### PR DESCRIPTION
**Pull request by IuliuNovac**
This is the second option for object detection. You can import you own custom model by replacing the public/yolov5 folder.

### **Disclaimer, I don't know JavaScript, i have no clue how this language works!!!**
I just managed to get it work, since i need this feature. Expect code not to meet the guidelines and might be buggy.

- Currently missing a way to switch between different type of models in AIObjectDetectionActions. I have no clue how to solve this, so i came up with an horrible temporary solution. I would like some help.
- Missing the ability to import the labels name of the yolov5 models. Right now it's defined in ObjectDetectorYOLO.ts as in variable names.
 `    private static names = ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
        'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow'....`
- Need a nice way to upload the model via the browser, again, i don't know JavaScript at all, so i would need help to implement that.
- Option for fine tuning/retraining while labeling would be very nice. Something as deploying to AWS sage maker for training or locally. I might be able to do that tho, but not the front end part.

Support all general variations of Yolov5, but missing a way to to chose between them.

> After the release of YOLOv5 v6, users could export tfjs models with its main branch:

```
git clone https://github.com/ultralytics/yolov5.git
python export.py --weights yolov5s.pt --include tfjs
```

### Pre-flight checklist

- [ ] Unit tests for all non-trivial changes
- [x] Tested locally
- [ ] Updated wiki

### CODE REVIEW

**Comment by IuliuNovac** [comment]
resolve #156

**Comment by SkalskiP** [comment]
Hello @IuliuNovac 👋  ! And thank you very much for your interest in our project. I must admit that it is a pretty good starting point for further work, but as you yourself noticed it is not yet ready to be included in the production version of MS.

* In my opinion we should add internal support for the YOLOv5 model trained on the COCO dataset, but allow users to upload their own model. 
* Because I would like to be able to freely change the code and at the same time I would like to keep a trace of the fact that it was you who originally added the changes I propose to change the target branch for that PR from develop to a specially created one dedicated to adding this feature. Your code would be added to this separate branch, as I would make the appropriate corrections and then your code and my code together would go into develop.

File: package.json [hunk 1]
```diff
@@ -7,10 +7,10 @@
     "@tensorflow-models/coco-ssd": "^2.2.2",
     "@tensorflow-models/posenet": "^2.2.2",
     "@tensorflow/tfjs": "^3.8.0",
-    "@tensorflow/tfjs-backend-cpu": "^3.9.0",
-    "@tensorflow/tfjs-backend-webgl": "^3.9.0",
-    "@tensorflow/tfjs-core": "^3.9.0",
-    "@tensorflow/tfjs-node": "^3.9.0",
+    "@tensorflow/tfjs-backend-cpu": "^3.8.0",
```
Comment (IuliuNovac) [inline package.json:10]: I noticed issues with different versions of tfjs, the 3.9.0 on other hand led to some issues. Could be that my project wasn't rebuild correctly.

File: src/ai/ObjectDetectorYOLO.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import {updateActiveLabelType} from '../store/labels/actionCreators';
+import {DetectedObject, ObjectDetection} from '@tensorflow-models/coco-ssd';
+import { AIModel } from '../data/enums/AIModel';
+
+export class ObjectDetectorYolov5 {
+    private static model: tf.GraphModel;
+    private static width = 640;
+    private static height = 640;
+    public static AIModel = AIModel.OBJECT_DETECTION;
+    private static names = ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
```
Comment (IuliuNovac) [inline src/ai/ObjectDetectorYOLO.ts:17]: This is not ok, it should be imported from a separate json file.

File: src/ai/ObjectDetectorYOLO.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                        score: score.toFixed(2)
+                    });
+        }
+        tf.dispose(results);
+        tf.engine().endScope();
+        if (callback) {
+            callback(predictions);
+        }
+    }
+}
```
Comment (IuliuNovac) [inline src/ai/ObjectDetectorYOLO.ts:90]: Requires refactoring, adding a confidence threshold a user can select would be nice. I was training the model, but in the first iterations the confidence of identifying my object was low. Unfortunately i was not able to change the confidence coefficient.

File: src/data/enums/AIModel.ts [hunk 4]
```diff
@@ -1,4 +1,5 @@
 export enum AIModel {
     OBJECT_DETECTION = "OBJECT_DETECTION",
-    POSE_DETECTION = "POSE_DETECTION"
+    POSE_DETECTION = "POSE_DETECTION",
+    OBJECT_DETECTION_YOLOv5 = "OBJECT_DETECTION_YOLOv5"
 }
```
Comment (IuliuNovac) [inline src/data/enums/AIModel.ts:5]: Probably Object_Detection and Pose Detection should have a separate enum  to chose between SSD, YOLO, etc.

File: src/logic/actions/AIObjectDetectionActions.ts [hunk 5]
```diff
@@ -25,18 +27,43 @@ export class AIObjectDetectionActions {
...
-        ObjectDetector.predict(image, (predictions: DetectedObject[]) => {
-            const suggestedLabelNames = AIObjectDetectionActions.extractNewSuggestedLabelNames(LabelsSelector.getLabelNames(), predictions);
-            const rejectedLabelNames = AISelector.getRejectedSuggestedLabelList();
-            const newlySuggestedNames = AIActions.excludeRejectedLabelNames(suggestedLabelNames, rejectedLabelNames);
-            if (newlySuggestedNames.length > 0) {
-                store.dispatch(updateSuggestedLabelList(newlySuggestedNames));
-                store.dispatch(updateActivePopupType(PopupWindowType.SUGGEST_LABEL_NAMES));
-            } else {
-                store.dispatch(updateActivePopupType(null));
+        switch (ObjectDetectorYolov5.AIModel) {
```
Comment (IuliuNovac) [inline src/logic/actions/AIObjectDetectionActions.ts:30]: This is the horrible quick fix i made, too chunky.  Also, i am storing the variable in the ObjectDetectorYolov5, that's horrible. I did not know how pass the value in AIModel to this object.

**Comment by IuliuNovac** [comment]
Hi @SkalskiP , that sounds quite good. By no means someone should use this branch unless they just need to run YOLOv5 alone. I used it to train a yolov5 model. The steps where simple:

1. Label a small number of images
2. Export the datasets and train.
3. Convert the model to tfjs model
4. Upload it into the public/yolov5
5. Repeat, while the prediction fro yolov5 are acceptable.

Note: You should not rely on the yolo model to assist for labeling, but rather evaluate the short comings. Since you can skip the images labeled right, while focusing on labeling the faulty predictions.

**Comment by SkalskiP** [comment]
Hi @IuliuNovac, that's exactly how I envision this feature. The only difference is that instead of`public/yolov5` I would like the user to be able to just load the model files into the editor. I'm already working on integration with YOLOv5 on a separate branch. I plan to divide the work into two phases:
- Adding general support for YOLOv5. The user will be able to use a model trained on COCO dataset.
- Allow the user to load a YOLOv5 model trained on any dataset.

**Pull request merged**